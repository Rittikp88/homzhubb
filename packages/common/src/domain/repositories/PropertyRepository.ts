import { BootstrapAppService } from '@homzhub/common/src/services/BootstrapAppService';
import { IApiClient } from '@homzhub/common/src/network/Interfaces';

const ENDPOINTS = {
  allDetail: (): string => 'https://jsonplaceholder.typicode.com/todos/1', // TODO: change the url once the api is ready
  getDetailsById: (id: string | number): string => 'https://jsonplaceholder.typicode.com/todos/1',
};

class PropertyRepository {
  private apiClient: IApiClient;

  public constructor() {
    this.apiClient = BootstrapAppService.clientInstance;
  }

  public getDetails = async (): Promise<void> => {
    return await this.apiClient.get(ENDPOINTS.allDetail());
  };

  public getDetailsById = async (payload: any): Promise<void> => {
    return await this.apiClient.get(ENDPOINTS.getDetailsById(payload));
  };
}

const propertyRepository = new PropertyRepository();
export { propertyRepository as PropertyRepository };
