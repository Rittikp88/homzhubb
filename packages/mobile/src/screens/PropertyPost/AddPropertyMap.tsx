import React from 'react';
import { View, StyleSheet } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import MapView, { LatLng, MapEvent, Marker, PROVIDER_GOOGLE } from 'react-native-maps';
import { Formik, FormikActions, FormikProps, FormikValues } from 'formik';
import { GooglePlacesService } from '@homzhub/common/src/services/GooglePlaces/GooglePlacesService';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import { Button, FormButton, FormTextInput, Label, Text, WithShadowView } from '@homzhub/common/src/components';
import { BottomSheet } from '@homzhub/mobile/src/components/molecules/BottomSheet';
import Header from '@homzhub/mobile/src/components/molecules/Header';
import { AppStackParamList } from '@homzhub/mobile/src/navigation/AppNavigator';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';

type Props = WithTranslation & NavigationScreenProps<AppStackParamList, ScreensKeys.AddProperty>;

interface IState {
  markerLatLng: LatLng;
  isVisible: boolean;
  location: {
    projectName: string;
    unitNo: number;
    blockNo: string;
  };
}

class AddPropertyMap extends React.PureComponent<Props, IState> {
  public constructor(props: Props) {
    super(props);
    const {
      route: { params },
    } = props;
    const { initialLatitude, initialLongitude, primaryTitle, secondaryTitle } = params;
    this.state = {
      markerLatLng: {
        latitude: initialLatitude,
        longitude: initialLongitude,
      },
      isVisible: false,
      location: {
        projectName: `${primaryTitle}, ${secondaryTitle}`,
        unitNo: 0,
        blockNo: '',
      },
    };
  }

  public render(): React.ReactNode {
    const { markerLatLng } = this.state;
    const { t } = this.props;

    return (
      <View style={styles.container}>
        {this.renderHeader()}
        <MapView
          provider={PROVIDER_GOOGLE}
          style={styles.mapView}
          initialRegion={{
            ...markerLatLng,
            latitudeDelta: 0.001,
            longitudeDelta: 0.001,
          }}
        >
          <Marker draggable coordinate={markerLatLng} onDragEnd={this.onMarkerDragEnd} />
        </MapView>
        <WithShadowView outerViewStyle={styles.shadowView}>
          <Button
            type="primary"
            title={t('setLocation')}
            containerStyle={styles.buttonStyle}
            onPress={this.onPressSetLocation}
          />
        </WithShadowView>
        {this.renderBottomSheet()}
      </View>
    );
  }

  private renderHeader = (): React.ReactNode => {
    const {
      t,
      route: { params },
    } = this.props;
    const { primaryTitle, secondaryTitle } = params;
    return (
      <>
        <Header
          icon="left-arrow"
          iconColor={theme.colors.white}
          onIconPress={this.onBackPress}
          isHeadingVisible
          title={t('common:location')}
          titleType="small"
          titleFontType="semiBold"
          titleStyle={styles.textColor}
          backgroundColor={theme.colors.primaryColor}
        />
        <View style={styles.titleContainer}>
          <Text type="small" textType="semiBold" style={[styles.titlePrimary, styles.textColor]}>
            {primaryTitle}
          </Text>
          <Label type="regular" textType="regular" style={[styles.titleSecondary, styles.textColor]}>
            {secondaryTitle}
          </Label>
        </View>
      </>
    );
  };

  private renderBottomSheet = (): React.ReactElement => {
    const { isVisible, location } = this.state;
    const { t } = this.props;
    const formData = { ...location };
    return (
      <BottomSheet visible={isVisible} onCloseSheet={this.onClose} headerTitle={t('locationDetails')} sheetHeight={400}>
        <Formik initialValues={formData} onSubmit={this.onSaveLocation}>
          {(formProps: FormikProps<FormikValues>): React.ReactNode => {
            return (
              <>
                <View style={styles.fieldsView}>
                  <FormTextInput
                    autoFocus
                    name="projectName"
                    label={t('projectName')}
                    inputType="default"
                    maxLength={100}
                    placeholder={t('projectName')}
                    formProps={formProps}
                  />
                  <View style={styles.contentView}>
                    <View style={styles.subContentView}>
                      <FormTextInput name="unitNo" label={t('unitNo')} inputType="number" formProps={formProps} />
                    </View>
                    <View style={styles.flexOne}>
                      <FormTextInput name="blockNo" label={t('blockNo')} inputType="default" formProps={formProps} />
                    </View>
                  </View>
                </View>
                <WithShadowView outerViewStyle={styles.shadowView}>
                  <FormButton
                    type="primary"
                    title={t('saveLocation')}
                    containerStyle={styles.buttonStyle}
                    // @ts-ignore
                    onPress={formProps.handleSubmit}
                    formProps={formProps}
                  />
                </WithShadowView>
              </>
            );
          }}
        </Formik>
      </BottomSheet>
    );
  };

  private onPressSetLocation = (): void => {
    const { isVisible } = this.state;
    this.setState({ isVisible: !isVisible });
  };

  private onClose = (): void => {
    this.setState({ isVisible: false });
  };

  private onBackPress = (): void => {
    const {
      navigation: { goBack },
    } = this.props;
    goBack();
  };

  private onMarkerDragEnd = (event: MapEvent): void => {
    const { latitude, longitude } = event.nativeEvent.coordinate;
    const {
      navigation: { setParams },
    } = this.props;

    GooglePlacesService.getLocationData({ lng: longitude, lat: latitude })
      .then((locData) => {
        const { formatted_address } = locData;
        const { location } = this.state;
        const { primaryAddress, secondaryAddress } = GooglePlacesService.getSplitAddress(formatted_address);
        setParams({
          primaryTitle: primaryAddress,
          secondaryTitle: secondaryAddress,
        });
        this.setState({
          location: {
            ...location,
            projectName: formatted_address,
          },
          markerLatLng: {
            longitude,
            latitude,
          },
        });
      })
      .catch((e: Error): void => {
        AlertHelper.error({ message: e.message });
      });
  };

  private onSaveLocation = (values: FormikValues, formActions: FormikActions<FormikValues>): void => {
    formActions.setSubmitting(true);
    // TODO: Add logic
  };
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  mapView: {
    flex: 1,
  },
  titleContainer: {
    backgroundColor: theme.colors.primaryColor,
    paddingStart: 16,
  },
  titlePrimary: {
    marginTop: 4,
    marginBottom: 8,
  },
  titleSecondary: {
    marginBottom: 12,
  },
  textColor: {
    color: theme.colors.white,
  },
  shadowView: {
    paddingTop: 10,
    marginBottom: PlatformUtils.isIOS() ? 20 : 0,
    paddingBottom: 0,
  },
  buttonStyle: {
    flex: 0,
    margin: 16,
  },
  fieldsView: {
    marginHorizontal: 24,
    marginBottom: 24,
  },
  contentView: {
    flexDirection: 'row',
  },
  subContentView: {
    flex: 1,
    marginRight: 16,
  },
  flexOne: {
    flex: 1,
  },
});

const HOC = withTranslation(LocaleConstants.namespacesKey.propertyPost)(AddPropertyMap);
export { HOC as AddPropertyMap };
