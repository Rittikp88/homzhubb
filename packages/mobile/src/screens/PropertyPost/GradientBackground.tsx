import React from 'react';
import { StatusBar, StyleSheet } from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import { theme } from '@homzhub/common/src/styles/theme';

interface IProps {
  children?: string | React.ReactNode;
}

export const GradientBackground = ({ children }: IProps): React.ReactElement => {
  return (
    <LinearGradient
      useAngle
      angle={54.15}
      colors={[theme.colors.propertyGradientA, theme.colors.propertyGradientB]}
      locations={[0.1955, 0.7738]}
      style={styles.container}
    >
      <StatusBar translucent backgroundColor="transparent" />
      {children}
    </LinearGradient>
  );
};

const styles = StyleSheet.create({
  container: {
    ...theme.globalStyles.center,
    flexDirection: 'row',
  },
});
