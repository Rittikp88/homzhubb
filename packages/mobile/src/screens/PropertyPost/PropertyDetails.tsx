import React from 'react';
import { StyleSheet, View, ScrollView } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { theme } from '@homzhub/common/src/styles/theme';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { PropertyActions } from '@homzhub/common/src/modules/property/actions';
import { PropertySelector } from '@homzhub/common/src/modules/property/selectors';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { Button, WithShadowView } from '@homzhub/common/src/components';
import Header from '@homzhub/mobile/src/components/molecules/Header';
import { PropertyDetailsLocation } from '@homzhub/mobile/src/components/molecules/PropertyDetailsLocation';
import PropertyDetailsItems from '@homzhub/mobile/src/components/organisms/PropertyDetailItems';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AppStackParamList } from '@homzhub/mobile/src/navigation/AppNavigator';

interface IDispatchProps {
  getPropertyDetails: () => void;
  getPropertyDetailsById: (id: string | number) => void;
}

interface IStateProps {
  property: any;
  spaceAvailable: any;
}

type libraryProps = WithTranslation & NavigationScreenProps<AppStackParamList, ScreensKeys.PropertyDetailsScreen>;
type Props = IDispatchProps & IStateProps & libraryProps;

interface IPropertyDetailsState {
  propertyGroupSelectedIndex: string | number;
  propertyGroupTypeSelectedIndex: string | number;
  isSelected: boolean;
}

class PropertyDetails extends React.PureComponent<Props, IPropertyDetailsState> {
  public state = {
    propertyGroupSelectedIndex: 0,
    propertyGroupTypeSelectedIndex: 0,
    isSelected: false,
  };

  public componentDidMount(): void {
    const { getPropertyDetails, getPropertyDetailsById } = this.props;
    const { propertyGroupSelectedIndex } = this.state;
    getPropertyDetails();
    // TODO: Move this to saga in get property details once the api is ready
    getPropertyDetailsById(propertyGroupSelectedIndex);
  }

  public render(): React.ReactNode {
    const { property, spaceAvailable, t } = this.props;
    const { propertyGroupSelectedIndex, propertyGroupTypeSelectedIndex, isSelected } = this.state;
    if (!property || !spaceAvailable) {
      return null;
    }
    return (
      <View style={styles.container}>
        <Header
          backgroundColor={theme.colors.primaryColor}
          icon="left-arrow"
          iconColor="white"
          onIconPress={this.handleIconPress}
          isHeadingVisible
          title={t('propertyDetails:headerTitle')}
          titleType="small"
          titleFontType="semiBold"
          titleStyle={styles.headerTitle}
        />
        <ScrollView style={styles.scrollContainer}>
          <PropertyDetailsLocation
            propertyName="Kalapataru Splendour"
            propertyAddress="Shankar Kalat Nagar, Maharashtra 411057"
            onNavigate={this.onNavigateToMaps}
          />
          <PropertyDetailsItems
            data={property}
            propertyGroupSelectedIndex={propertyGroupSelectedIndex}
            propertyGroupTypeSelectedIndex={propertyGroupTypeSelectedIndex}
            spaceAvailable={spaceAvailable ?? []}
            onPropertyGroupChange={this.onPropertyGroupChange}
            onPropertyGroupTypeChange={this.onPropertyGroupTypeChange}
          />
        </ScrollView>
        <WithShadowView outerViewStyle={styles.shadowView}>
          <Button
            type="primary"
            title={t('common:submit')}
            disabled={!isSelected}
            containerStyle={styles.buttonStyle}
            onPress={this.onSubmit}
          />
        </WithShadowView>
      </View>
    );
  }

  public onNavigateToMaps = (): void => {};

  public onPropertyGroupChange = (index: string | number): void => {
    const { getPropertyDetailsById } = this.props;
    this.setState({
      propertyGroupSelectedIndex: index,
      propertyGroupTypeSelectedIndex: 0,
    });
    getPropertyDetailsById(index);
  };

  public onPropertyGroupTypeChange = (index: string | number): void => {
    this.setState({
      propertyGroupTypeSelectedIndex: index,
    });
  };

  private onSubmit = (): void => {
    // Add logic
  };

  public handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };
}

const mapStateToProps = (state: IState): IStateProps => {
  return {
    property: PropertySelector.getPropertyDetails(state),
    spaceAvailable: PropertySelector.getPropertySpaceAvailable(state),
  };
};

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { getPropertyDetails, getPropertyDetailsById } = PropertyActions;
  return bindActionCreators(
    {
      getPropertyDetails,
      getPropertyDetailsById,
    },
    dispatch
  );
};

export default connect<IStateProps, IDispatchProps, WithTranslation, IState>(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation()(PropertyDetails));

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  scrollContainer: {
    flex: 1,
  },
  headerTitle: {
    color: theme.colors.white,
  },
  shadowView: {
    paddingTop: 10,
    marginBottom: PlatformUtils.isIOS() ? 20 : 0,
    paddingBottom: 0,
  },
  buttonStyle: {
    flex: 0,
    margin: 16,
  },
});
