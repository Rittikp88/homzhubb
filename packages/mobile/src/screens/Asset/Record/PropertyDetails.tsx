import React from 'react';
import { StyleSheet, View, ScrollView } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { CommonRepository } from '@homzhub/common/src/domain/repositories/CommonRepository';
import { GooglePlacesService } from '@homzhub/common/src/services/GooglePlaces/GooglePlacesService';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { PropertyActions } from '@homzhub/common/src/modules/property/actions';
import { PropertySelector } from '@homzhub/common/src/modules/property/selectors';
import {
  IUpdateAssetDetails,
  SpaceAvailableTypes,
  ISpaceAvailable,
  ISpaceAvailablePayload,
} from '@homzhub/common/src/domain/repositories/interfaces';
import { theme } from '@homzhub/common/src/styles/theme';
import { icons } from '@homzhub/common/src/assets/icon';
import { Button, IDropdownOption, WithShadowView } from '@homzhub/common/src/components';
import { Header, PropertyDetailsLocation, StateAwareComponent } from '@homzhub/mobile/src/components';
import PropertyDetailsItems from '@homzhub/mobile/src/components/organisms/PropertyDetailItems';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { PropertyPostStackParamList } from '@homzhub/mobile/src/navigation/PropertyPostStack';
import { IPropertyDetailsData, IPropertyTypes } from '@homzhub/common/src/domain/models/Property';

interface IDispatchProps {
  getPropertyDetails: () => void;
}

interface IStateProps {
  propertyId: number;
  property: IPropertyDetailsData[] | null;
  isLoading: boolean;
}

type libraryProps = NavigationScreenProps<PropertyPostStackParamList, ScreensKeys.PropertyDetailsScreen>;
type Props = IDispatchProps & IStateProps & libraryProps & WithTranslation;

interface IPropertyDetailsState {
  projectName: string;
  propertyGroupSelectedIndex: string | number;
  propertyGroupTypeSelectedIndex: string | number;
  areaUnits: IDropdownOption[];
  spaceAvailable: ISpaceAvailable;
  carpetAreaError: boolean;
}

export class PropertyDetails extends React.PureComponent<Props, IPropertyDetailsState> {
  public state = {
    projectName: '',
    propertyGroupSelectedIndex: 0,
    propertyGroupTypeSelectedIndex: 0,
    areaUnits: [],
    spaceAvailable: {
      bedroom: 0,
      bathroom: 0,
      balcony: 0,
      floorNumber: 0,
      totalFloors: 0,
      carpetArea: '',
      areaUnit: 'SQ_FT',
    },
    carpetAreaError: false,
  };

  public componentDidMount = async (): Promise<void> => {
    const { getPropertyDetails } = this.props;
    getPropertyDetails();
    await this.getProjectName();
    await this.getCarpetAreaUnits();
  };

  public render(): React.ReactNode {
    const { isLoading } = this.props;
    return (
      <StateAwareComponent loading={isLoading} renderComponent={this.renderComponentContent()} testID="stateAware" />
    );
  }

  public renderComponentContent = (): React.ReactElement | null => {
    const { property, t } = this.props;
    const {
      propertyGroupSelectedIndex,
      propertyGroupTypeSelectedIndex,
      areaUnits,
      spaceAvailable,
      carpetAreaError,
      projectName,
    } = this.state;
    if (!property) {
      return null;
    }
    const address = GooglePlacesService.getSplitAddress(projectName);
    return (
      <View style={styles.container}>
        <Header
          type="primary"
          icon={icons.leftArrow}
          onIconPress={this.handleIconPress}
          isHeadingVisible
          title={t('property:headerTitle')}
          testID="headerView"
        />
        <ScrollView style={styles.scrollContainer}>
          <PropertyDetailsLocation
            propertyName={address.primaryAddress}
            propertyAddress={address.secondaryAddress}
            onNavigate={this.onNavigateToMaps}
            testID="propertyLocation"
          />
          <PropertyDetailsItems
            data={property}
            areaUnits={areaUnits}
            spaceAvailable={spaceAvailable}
            propertyGroupSelectedIndex={propertyGroupSelectedIndex}
            propertyGroupTypeSelectedIndex={propertyGroupTypeSelectedIndex}
            carpetAreaError={carpetAreaError}
            onPropertyGroupChange={this.onPropertyGroupChange}
            onPropertyGroupTypeChange={this.onPropertyGroupTypeChange}
            onSpaceAvailableValueChange={this.onSpaceAvailableValueChange}
            onCommercialPropertyChange={this.updateSpaceAvailable}
          />
        </ScrollView>
        <WithShadowView outerViewStyle={styles.shadowView}>
          <Button
            type="primary"
            title={t('common:submit')}
            disabled={!spaceAvailable.carpetArea}
            containerStyle={styles.buttonStyle}
            onPress={this.onSubmit}
          />
        </WithShadowView>
      </View>
    );
  };

  public onNavigateToMaps = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  public onPropertyGroupChange = (index: string | number): void => {
    const { spaceAvailable } = this.state;
    this.setState({
      propertyGroupSelectedIndex: index,
      propertyGroupTypeSelectedIndex: 0,
      spaceAvailable: {
        ...spaceAvailable,
        bathroom: 0,
        carpetArea: '',
      },
    });
  };

  public onPropertyGroupTypeChange = (index: string | number): void => {
    this.setState({
      propertyGroupTypeSelectedIndex: index,
    });
  };

  public onSpaceAvailableValueChange = (item: IPropertyTypes, index: string | number): void => {
    switch (item.name) {
      case SpaceAvailableTypes.BATHROOM:
        this.updateSpaceAvailable('bathroom', index);
        break;
      case SpaceAvailableTypes.BALCONY:
        this.updateSpaceAvailable('balcony', index);
        break;
      case SpaceAvailableTypes.TOTAL_FLOORS:
        this.updateSpaceAvailable('totalFloors', index);
        break;
      case SpaceAvailableTypes.FLOOR_NUMBER:
        this.updateSpaceAvailable('floorNumber', index);
        break;
      case SpaceAvailableTypes.BEDROOM:
        this.updateSpaceAvailable('bedroom', index);
        break;
      default:
        break;
    }
  };

  private onSubmit = async (): Promise<void> => {
    const { navigation, property, propertyId } = this.props;
    const {
      spaceAvailable: { carpetArea, areaUnit, floorNumber, totalFloors },
      propertyGroupSelectedIndex,
      propertyGroupTypeSelectedIndex,
    } = this.state;

    if (!property) {
      return;
    }
    const spaces = this.getSpacesForProperty();
    try {
      let requestPayload: IUpdateAssetDetails = {
        asset_type: Number(property[propertyGroupSelectedIndex].asset_types[propertyGroupTypeSelectedIndex].id),
        spaces,
        carpet_area: carpetArea.toString(),
        carpet_area_unit: areaUnit,
      };
      if (!carpetArea) {
        this.setState({ carpetAreaError: true });
      }
      if (propertyGroupSelectedIndex === 1) {
        requestPayload = {
          ...requestPayload,
          floor_number: floorNumber,
          total_floors: totalFloors,
        };
      }
      await AssetRepository.updateAsset(propertyId, requestPayload);
      navigation.navigate(ScreensKeys.AssetPropertyImages);
    } catch (e) {
      const error = ErrorUtils.getErrorMessage(e.details);
      AlertHelper.error({ message: error });
    }
  };

  public getSpacesForProperty(): ISpaceAvailablePayload[] {
    const {
      spaceAvailable: { bathroom, bedroom, balcony },
      propertyGroupSelectedIndex,
    } = this.state;
    const { property } = this.props;
    const spaces: ISpaceAvailablePayload[] = [];
    const countOfSpace = (type: string): number => {
      switch (type) {
        case SpaceAvailableTypes.BATHROOM:
          return bathroom;
        case SpaceAvailableTypes.BEDROOM:
          return bedroom;
        case SpaceAvailableTypes.BALCONY:
          return balcony;
        default:
          return 0;
      }
    };
    property?.[propertyGroupSelectedIndex].space_types.forEach((spaceType) => {
      spaces.push({
        space_type: spaceType.id,
        count: countOfSpace(spaceType.name),
      });
    });
    return spaces;
  }

  public getCarpetAreaUnits = async (): Promise<void> => {
    try {
      const response = await CommonRepository.getCarpetAreaUnits();
      this.setState({
        areaUnits: response,
      });
    } catch (e) {
      const error = ErrorUtils.getErrorMessage(e.details);
      AlertHelper.error({ message: error });
    }
  };

  public handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  public updateSpaceAvailable = (type: string, value: string | number): void => {
    const { spaceAvailable } = this.state;
    this.setState({
      spaceAvailable: {
        ...spaceAvailable,
        [type]: value,
      },
      carpetAreaError: false,
    });
  };

  private getProjectName = async (): Promise<void> => {
    const { propertyId } = this.props;
    try {
      const response = await AssetRepository.getAssetById(propertyId);
      this.setState({ projectName: response.projectName });
    } catch (e) {
      const error = ErrorUtils.getErrorMessage(e.details);
      AlertHelper.error({ message: error });
    }
  };
}

export const mapStateToProps = (state: IState): IStateProps => {
  const { getCurrentPropertyId, getPropertyDetails, getPropertyLoadingState } = PropertySelector;
  return {
    property: getPropertyDetails(state),
    propertyId: getCurrentPropertyId(state),
    isLoading: getPropertyLoadingState(state),
  };
};

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { getPropertyDetails } = PropertyActions;
  return bindActionCreators(
    {
      getPropertyDetails,
    },
    dispatch
  );
};

export default connect<IStateProps, IDispatchProps, WithTranslation, IState>(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation()(PropertyDetails));

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  scrollContainer: {
    flex: 1,
  },
  shadowView: {
    paddingTop: 10,
    marginBottom: PlatformUtils.isIOS() ? 20 : 0,
    paddingBottom: 0,
  },
  buttonStyle: {
    flex: 0,
    margin: 16,
  },
});
