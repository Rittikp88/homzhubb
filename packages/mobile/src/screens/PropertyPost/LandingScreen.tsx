import React from 'react';
import { StyleSheet, View, Image } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { IUser } from '@homzhub/common/src/domain/models/User';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { IRefreshTokenPayload, IUserPayload } from '@homzhub/common/src/domain/repositories/interfaces';
import { StorageService, StorageKeys } from '@homzhub/common/src/services/storage/StorageService';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { theme } from '@homzhub/common/src/styles/theme';
import { images } from '@homzhub/common/src/assets/images';
import { Text, Label, Button, SVGUri } from '@homzhub/common/src/components';
import { GradientBackground } from '@homzhub/mobile/src/components/molecules/GradientBackground';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AppStackParamList } from '@homzhub/mobile/src/navigation/AppNavigator';

interface IStateProps {
  user: IUser | null;
}

interface IDispatchProps {
  logout: (data: IRefreshTokenPayload) => void;
}

type libraryProps = WithTranslation & NavigationScreenProps<AppStackParamList, ScreensKeys.PropertyPostLandingScreen>;
type Props = IStateProps & IDispatchProps & libraryProps;

export class LandingScreen extends React.PureComponent<Props, {}> {
  public render(): React.ReactNode {
    const { t, user } = this.props;
    return (
      <GradientBackground>
        <View style={styles.container}>
          <View style={styles.imagesContainer}>
            <Image source={images.landingScreenLogo} />
            <Text type="regular" textType="semiBold">
              {t('welcomeUser', { username: user?.full_name ?? '' })}
            </Text>
            <Text type="small" textType="regular" style={styles.description}>
              {t('description')}
            </Text>
            <SVGUri
              width={295}
              height={133}
              viewBox="0 0 295 133"
              preserveAspectRatio="xMidYMid meet"
              uri="https://homzhub-bucket.s3.amazonaws.com/Group+1168.svg"
            />
            <Label type="regular" textType="regular">
              {t('searchProject')}
            </Label>
          </View>
          <View style={styles.buttonContainer}>
            <Button
              title={t('addProperty')}
              type="primary"
              onPress={this.onAddProperty}
              containerStyle={styles.addProperty}
              testID="btnAddProperty"
            />
            <Label type="large" textType="regular" style={styles.logoutContainer}>
              {t('logoutHelperText')} &nbsp;
              <Label type="large" textType="bold" style={styles.logout} onPress={this.logout} testID="lblLogout">
                {t('common:logout')}
              </Label>
            </Label>
          </View>
        </View>
      </GradientBackground>
    );
  }

  public onAddProperty = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.SearchPropertyOwner);
  };

  public logout = async (): Promise<void> => {
    const { logout } = this.props;
    const user: IUserPayload | null = (await StorageService.get(StorageKeys.USER)) ?? null;
    if (!user) {
      return;
    }
    const { refresh_token } = user;
    const logoutPayload = {
      refresh_token,
    };
    logout(logoutPayload);
  };
}

export const mapStateToProps = (state: IState): IStateProps => {
  return {
    user: UserSelector.getUserDetails(state),
  };
};

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { logout } = UserActions;
  return bindActionCreators(
    {
      logout,
    },
    dispatch
  );
};

export default connect<IStateProps, IDispatchProps, WithTranslation, IState>(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.property)(LandingScreen));

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
    shadowColor: theme.colors.shadow,
    shadowOffset: {
      width: 0,
      height: 10,
    },
    shadowOpacity: 0.51,
    shadowRadius: 13.16,
    elevation: 20,
    width: theme.viewport.width,
    maxHeight: theme.viewport.height - 100,
    borderRadius: 8,
    margin: theme.layout.screenPadding,
    paddingHorizontal: theme.layout.screenPadding,
  },
  imagesContainer: {
    justifyContent: 'space-around',
    alignItems: 'center',
    flex: 0.9,
  },
  description: {
    textAlign: 'center',
  },
  buttonContainer: {
    justifyContent: 'space-around',
    alignItems: 'stretch',
    marginVertical: 20,
  },
  addProperty: {
    flex: 0,
  },
  logoutContainer: {
    alignSelf: 'center',
    marginVertical: 10,
  },
  logout: {
    color: theme.colors.blue,
  },
});
