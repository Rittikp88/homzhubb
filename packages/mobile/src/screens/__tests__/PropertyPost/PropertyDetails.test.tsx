import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PropertyAssetGroupData, ResidentialPropertyTypeData } from '@homzhub/common/src/mocks/PropertyDetails';
import { PropertyDetails } from '@homzhub/mobile/src/screens/PropertyPost/PropertyDetails';
import { PropertyRepository } from '@homzhub/common/src/domain/repositories/PropertyRepository';

const mock = jest.fn();

describe('Property Details Screen Component', () => {
  let component: ShallowWrapper;
  let props: any;

  beforeEach(() => {
    props = {
      getPropertyDetails: jest.fn(),
      getPropertyDetailsById: jest.fn(),
      property: PropertyAssetGroupData,
      spaceAvailable: ResidentialPropertyTypeData,
      navigation: {
        navigate: mock,
        goBack: mock,
      },
    };
    component = shallow(<PropertyDetails {...props} t={(key: string): string => key} />);
  });

  it('should render property details screen', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should navigate to previos screen', () => {
    // @ts-ignore
    component.find('[testID="propertyLocation"]').prop('onNavigate')();
    expect(mock).toHaveBeenCalled();
  });

  it('should call the project name', async () => {
    const data = {
      project_name: 'Project A',
      unit_number: '12',
      block_number: 'A',
      latitude: '109.12',
      longitude: '110.12',
      carpet_area: '1000',
      carpet_area_unit: 'Sq.ft',
      floor_number: 1,
      total_floors: 2,
      asset_type: 0,
    };
    jest.spyOn(PropertyRepository, 'getAssetById').mockImplementation(() => Promise.resolve(data));
    const response = await PropertyRepository.getAssetById(1);
    component.setState({ projectName: response.project_name });
    expect(component.state('projectName')).toStrictEqual(data.project_name);
  });
});
