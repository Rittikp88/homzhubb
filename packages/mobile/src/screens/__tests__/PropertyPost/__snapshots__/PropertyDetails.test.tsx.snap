// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property Details Screen Component should render property details screen 1`] = `
<Connect(withI18nextTranslation(PropertyDetails))
  getPropertyDetails={[MockFunction]}
  getPropertyDetailsById={[MockFunction]}
  navigation={
    Object {
      "navigate": [MockFunction],
    }
  }
  property={
    Array [
      Object {
        "id": 1,
        "name": "Residential",
        "property_types": Array [
          Object {
            "id": 1,
            "name": "Residential House",
          },
          Object {
            "id": 2,
            "name": "Apartment/Condo",
          },
          Object {
            "id": 3,
            "name": "Villa",
          },
          Object {
            "id": 4,
            "name": "Penthouse",
          },
          Object {
            "id": 5,
            "name": "Studio Apartment",
          },
          Object {
            "id": 6,
            "name": "Independent Floor",
          },
        ],
      },
      Object {
        "id": 2,
        "name": "Commercial",
        "property_types": Array [
          Object {
            "id": 1,
            "name": "Commercial Office Space",
          },
          Object {
            "id": 2,
            "name": "Office in IT park/SEZ",
          },
          Object {
            "id": 3,
            "name": "Commercial Showroom",
          },
          Object {
            "id": 4,
            "name": "Warehouse/Godown",
          },
        ],
      },
    ]
  }
  spaceAvailable={
    Array [
      Object {
        "id": 1,
        "name": "Room",
      },
      Object {
        "id": 2,
        "name": "Balcony",
      },
      Object {
        "id": 3,
        "name": "Bathroom",
      },
    ]
  }
/>
`;
